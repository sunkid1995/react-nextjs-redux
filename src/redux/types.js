export const PREFIXES = {
  AUTH: 'auth/',
  SERVICE_API: 'serviceApi/',
  SHARED: 'shared/',
};

export const INIT_ACTION = { type: '@@redux/INIT' };

export const AUTH = {
  AUTHORIZE: 'auth/AUTHORIZE',
  DEAUTHORIZE: 'auth/DEAUTHORIZE',
};

export const SERVICE_API = {
  AUTH_WITH_EMAIL: 'serviceApi/AUTH_WITH_EMAIL',
  FORGOT_PASSWORD: 'serviceApi/FORGOT_PASSWORD',
  RESET_PASSWORD: 'serviceApi/RESET_PASSWORD',
};

export const SHARED = {

};
