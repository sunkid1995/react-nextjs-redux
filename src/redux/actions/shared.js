// import { createAction } from 'redux-actions';

// // Redux
// import { SHARED } from 'src/redux/types';
