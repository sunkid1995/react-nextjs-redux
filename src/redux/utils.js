import _ from 'lodash';

import { IN_DEV_MODE } from 'src/constants';

export function buildErrorMessage(errors) {
  if (_.isEmpty(errors)) return null;

  const transformer = ({ code, message }) => (
    IN_DEV_MODE && code != null ? `${code} - ${message}` : message
  );
  return _.compact(_.map(errors, transformer)).join(', ');
}

export function buildHeaders(state) {
  const { auth: { user: { auth_token: token } = {} } } = state;
  const bearerToken = token != null ? `Bearer ${token}` : undefined;

  return {
    Accept: 'application/json',
    Authorization: bearerToken,
  };
}

export function responseInterceptor(response, callback) {
  if (response == null) return;
  return callback(response);
}
