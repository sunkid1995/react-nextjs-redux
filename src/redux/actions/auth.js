import { createAction } from 'redux-actions';

// Redux
import { AUTH } from 'src/redux/types';

export const authorize = createAction(AUTH.AUTHORIZE);
export const deauthorize = createAction(AUTH.DEAUTHORIZE);
