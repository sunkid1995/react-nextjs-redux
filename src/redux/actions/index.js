import * as auth from './auth';
import * as serviceApi from './serviceApi';
import * as shared from './shared';

export { auth, serviceApi, shared };
