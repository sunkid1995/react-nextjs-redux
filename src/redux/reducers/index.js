import auth from './auth';
import meta from './meta';
import serviceApi from './serviceApi';
import shared from './shared';

export { auth, meta, serviceApi, shared };
