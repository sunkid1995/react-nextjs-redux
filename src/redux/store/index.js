import { applyMiddleware, compose, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { persistCombineReducers, persistStore } from 'redux-persist';
import reduxThunk from 'redux-thunk';

import { IN_DEV_MODE } from 'src/constants';
import { auth, meta, serviceApi, shared } from 'src/redux/reducers';

// Locals
import buildAxiosMiddleware from './buildAxiosMiddleware';
import buildPersistorConfigs from './buildPersistorConfigs';

const composeEnhancer = IN_DEV_MODE ? composeWithDevTools({}) : compose;

export default function makeStore(/* initialState, options */) {
  const configs = buildPersistorConfigs();

  const store = createStore(
    persistCombineReducers(configs, { auth, meta, serviceApi, shared }),
    composeEnhancer(applyMiddleware(buildAxiosMiddleware(), reduxThunk)),
  );

  persistStore(store);
  return store;
}
