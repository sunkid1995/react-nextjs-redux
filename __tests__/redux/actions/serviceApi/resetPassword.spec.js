import _ from 'lodash';

// Helpers
import { buildErrorAction, buildSuccessAction, mockAxios } from '__tests__/helpers/mockApiRequests';
import buildMockStore from '__tests__/helpers/buildMockStore';

// Redux
import { serviceApi } from 'src/redux/actions';

describe('redux/actions/serviceApi/resetPassword', () => {
  const store = buildMockStore({ authorized: true });

  const token = 'reset_password_token_xxx';
  const password = '12345678';
  const passwordConfirm = password;
  const url = '/users/password';

  const requestAction = {
    type: 'serviceApi/RESET_PASSWORD',
    payload: {
      dataKey: 'passwordReset',
      manifest: { reserved: true },
      request: {
        url,
        data: {
          password,
          reset_password_token: token,
          password_confirmation: passwordConfirm,
        },
        method: 'patch',
      },
    },
  };
  const successAction = buildSuccessAction(requestAction, {});
  const { errorAction, mockErrorResponse } = buildErrorAction(requestAction);

  /* eslint-disable newline-per-chained-call */
  mockAxios
    .onPatch(url).replyOnce(200, {})
    .onPatch(url).replyOnce(...mockErrorResponse);
  /* eslint-enable newline-per-chained-call */

  beforeEach(() => store.clearActions());

  test('success', async () => {
    await store.dispatch(serviceApi.resetPassword({ password, passwordConfirm, token }));
    const actions = store.getActions();
    expect(actions.length).toEqual(2);
    expect(actions).toEqual([requestAction, successAction]);
    expect(actions).toMatchSnapshot();

    const { payload: { data } } = _.last(actions);
    expect(data).toEqual({});
  });

  test('error', async () => {
    await store.dispatch(serviceApi.resetPassword({ password, passwordConfirm, token }));
    const actions = store.getActions();
    expect(actions.length).toEqual(2);
    expect(actions).toEqual([requestAction, errorAction]);
    expect(actions).toMatchSnapshot();
  });
});
