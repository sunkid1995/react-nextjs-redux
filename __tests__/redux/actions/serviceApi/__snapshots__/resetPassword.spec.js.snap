// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/actions/serviceApi/resetPassword error 1`] = `
Array [
  Object {
    "payload": Object {
      "dataKey": "passwordReset",
      "manifest": Object {
        "reserved": true,
      },
      "request": Object {
        "data": Object {
          "password": "12345678",
          "password_confirmation": "12345678",
          "reset_password_token": "reset_password_token_xxx",
        },
        "method": "patch",
        "url": "/users/password",
      },
    },
    "type": "serviceApi/RESET_PASSWORD",
  },
  Object {
    "error": Object {
      "message": "serviceApi/RESET_PASSWORD request failed",
      "response": Object {
        "data": Object {
          "errors": Array [
            Object {
              "message": "serviceApi/RESET_PASSWORD request failed",
            },
          ],
        },
        "status": 500,
      },
    },
    "meta": Object {
      "previousAction": Object {
        "payload": Object {
          "dataKey": "passwordReset",
          "manifest": Object {
            "reserved": true,
          },
          "request": Object {
            "data": Object {
              "password": "12345678",
              "password_confirmation": "12345678",
              "reset_password_token": "reset_password_token_xxx",
            },
            "method": "patch",
            "url": "/users/password",
          },
        },
        "type": "serviceApi/RESET_PASSWORD",
      },
    },
    "type": "serviceApi/RESET_PASSWORD:ERROR",
  },
]
`;

exports[`redux/actions/serviceApi/resetPassword success 1`] = `
Array [
  Object {
    "payload": Object {
      "dataKey": "passwordReset",
      "manifest": Object {
        "reserved": true,
      },
      "request": Object {
        "data": Object {
          "password": "12345678",
          "password_confirmation": "12345678",
          "reset_password_token": "reset_password_token_xxx",
        },
        "method": "patch",
        "url": "/users/password",
      },
    },
    "type": "serviceApi/RESET_PASSWORD",
  },
  Object {
    "meta": Object {
      "previousAction": Object {
        "payload": Object {
          "dataKey": "passwordReset",
          "manifest": Object {
            "reserved": true,
          },
          "request": Object {
            "data": Object {
              "password": "12345678",
              "password_confirmation": "12345678",
              "reset_password_token": "reset_password_token_xxx",
            },
            "method": "patch",
            "url": "/users/password",
          },
        },
        "type": "serviceApi/RESET_PASSWORD",
      },
    },
    "payload": Object {
      "data": Object {},
    },
    "type": "serviceApi/RESET_PASSWORD:SUCCESS",
  },
]
`;
